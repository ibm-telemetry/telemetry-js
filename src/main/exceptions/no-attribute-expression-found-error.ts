/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
/**
 * Exception thrown when  JsxElement's attribute has an undefined expression value.
 */
export class NoAttributeExpressionFoundError extends Error {
  constructor(attr: string) {
    super('No expression found in attribute ' + attr)
  }
}
