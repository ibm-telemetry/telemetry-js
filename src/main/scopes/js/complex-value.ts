/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Object representing a complex value.
 */
export class ComplexValue {
  constructor(public readonly complexValue: unknown) {}
}
