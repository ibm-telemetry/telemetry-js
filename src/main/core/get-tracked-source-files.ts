/*
 * Copyright IBM Corp. 2023, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { readFile } from 'node:fs/promises'
import * as path from 'node:path'

import * as ts from 'typescript'

import { type Logger } from './log/logger.js'
import { TrackedFileEnumerator } from './tracked-file-enumerator.js'
import { parseDocument } from 'htmlparser2'
import type { Document as HtmlDocument } from 'domhandler'
import { HtmlParsedFile } from '../scopes/wc/interfaces.js'

/**
 * Gets all tracked source files to consider for data collection,
 * filtered by supplied file extension array.
 *
 * @param cwd - Working directory for underlying command execution.
 * @param root - Root directory in which to search for tracked source files. This is an absolute
 * path.
 * @param logger - Logger instance to use.
 * @param fileExtensions - List of file extensions to filter files by.
 * @returns An array of source file objects.
 */
export async function getTrackedSourceFiles(
  cwd: string,
  root: string,
  logger: Logger,
  fileExtensions: string[]
) {
  logger.traceEnter('', 'getTrackedSourceFiles', [root, fileExtensions])

  const fileEnumerator = new TrackedFileEnumerator(logger)
  const files = []

  // If a file is passed instead of a directory, avoid the `git ls-tree` call
  if (fileExtensions.includes(path.extname(root))) {
    files.push(root)
  } else {
    files.push(
      ...(await fileEnumerator.find(cwd, root, (file) =>
        fileExtensions.includes(path.extname(file))
      ))
    )
  }

  const promises = files.map(async (file) => {
    return {
      fileName: file,
      async createSourceFile(): Promise<ts.SourceFile | HtmlDocument> {
        // Parse html file with `htmlparser2
        if (file.endsWith('html') || file.endsWith('htm')) {
          const parsedFile = parseDocument((await readFile(file)).toString()) as HtmlParsedFile
          parsedFile.fileName = file
        }

        // Parse any other files
        return ts.createSourceFile(
          file,
          (await readFile(file)).toString(),
          ts.ScriptTarget.ES2021,
          /* setParentNodes */ true
        )
      }
    }
  })

  const results = await Promise.all(promises)

  logger.traceExit(
    '',
    'getTrackedSourceFiles',
    results.map((result) => result.fileName)
  )
  return results
}
