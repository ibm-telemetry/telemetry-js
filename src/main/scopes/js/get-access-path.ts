/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as ts from 'typescript'

import type { Logger } from '../../core/log/logger.js'
import { ComplexValue } from './complex-value.js'
import { getNodeValueHandler } from './node-value-handler-map.js'

/**
 * Constructs a JsToken object from a given Identifier type AST node.
 *
 * @param node - TS node to retrieve access path for.
 * @param sourceFile - Root AST node.
 * @param logger - A logger instance.
 * @param currAccessPath - For internal use only, tracks current constructed access path.
 * @param topLevel - For internal use only, tracks top level function call.
 * @returns Access path to node represented as string array (chunks).
 */
export default function getAccessPath(
  node: ts.Node,
  sourceFile: ts.SourceFile,
  logger: Logger,
  currAccessPath: string[] = [],
  topLevel: boolean = true
): string[] {
  switch (node.kind) {
    case ts.SyntaxKind.Identifier:
      return [...currAccessPath, (node as ts.Identifier).escapedText.toString()]
    case ts.SyntaxKind.PropertyAccessExpression:
      currAccessPath.push((node as ts.PropertyAccessExpression).name.escapedText.toString())
      break
    case ts.SyntaxKind.ElementAccessExpression: {
      const argumentExpression = (node as ts.ElementAccessExpression).argumentExpression
      const data = getNodeValueHandler(argumentExpression.kind, sourceFile, logger).getData(
        argumentExpression
      )
      if (data instanceof ComplexValue) {
        currAccessPath = [
          ...currAccessPath,
          getAccessPath(argumentExpression, sourceFile, logger, [], false).reverse().join('.')
        ]
      } else {
        currAccessPath.push(data?.toString() ?? '')
      }
      break
    }
  }

  let accessPath = currAccessPath

  if ('expression' in node) {
    accessPath = getAccessPath(
      node.expression as ts.Node,
      sourceFile,
      logger,
      currAccessPath,
      false
    )
  }

  return topLevel ? accessPath.reverse() : accessPath
}
