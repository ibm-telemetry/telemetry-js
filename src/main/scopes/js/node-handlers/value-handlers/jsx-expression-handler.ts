/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type * as ts from 'typescript'

import { NoAttributeExpressionFoundError } from '../../../../exceptions/no-attribute-expression-found-error.js'
import { getNodeValueHandler } from '../../node-value-handler-map.js'
import { NodeValueHandler } from './node-value-handler.js'

/**
 * Holds logic to extract data from an AST node that is a JsxExpression kind.
 *
 */
export class JsxExpressionHandler extends NodeValueHandler {
  /**
   * Extracts string value of node.
   *
   * @param node - JsxExpression node to extract data from.
   * @returns Text value of node.
   * @throws NoAttributeExpressionFoundError if node's expression is undefined.
   */
  public getData(node: ts.JsxExpression) {
    if (node.expression === undefined) {
      throw new NoAttributeExpressionFoundError(node.getText(this.sourceFile))
    }

    return getNodeValueHandler(node.expression.kind, this.sourceFile, this.logger).getData(
      node.expression
    )
  }
}
