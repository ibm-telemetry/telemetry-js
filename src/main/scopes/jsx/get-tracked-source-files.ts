/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import path from 'node:path'

import * as ts from 'typescript'

import { type Logger } from '../../core/log/logger.js'
import { TrackedFileEnumerator } from '../../core/tracked-file-enumerator.js'

/**
 * Gets all tracked source files to consider for data collection.
 *
 * @param root - Root directory in which to search for tracked source files.
 * @param logger - Logger instance to use.
 * @returns An array of source file objects.
 */
export async function getTrackedSourceFiles(root: string, logger: Logger) {
  const fileEnumerator = new TrackedFileEnumerator(logger)
  const allowedSuffixes = ['js', 'jsx', 'ts', 'tsx']

  const files = await fileEnumerator.find(root, (file) =>
    allowedSuffixes.includes(path.extname(file))
  )

  const program = ts.createProgram(files, {})

  return program.getSourceFiles().filter((file) => !file.isDeclarationFile)
}
