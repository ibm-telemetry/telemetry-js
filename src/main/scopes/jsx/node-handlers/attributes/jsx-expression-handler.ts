/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type * as ts from 'typescript'

import { getNodeHandler } from '../../attributes-node-handler-map.js'
import { type ASTNodeHandler } from '../../interfaces.js'
import { DefaultHandler } from './default-handler.js'

/**
 * Holds logic to extract data from an AST node that is a JsxExpression kind.
 *
 */
export class JsxExpressionHandler implements ASTNodeHandler<ts.SyntaxKind.JsxExpression> {
  /**
   * Extracts string value of node.
   *
   * @param node - JsxExpression node to extract data from.
   * @param rootNode - FileSource root node that contains the supplied node.
   * @returns Text value of node.
   */
  public getData(node: ts.Node, rootNode: ts.SourceFile) {
    const nodeAsJsxExpression = node as ts.JsxExpression
    if (nodeAsJsxExpression.expression !== null && nodeAsJsxExpression.expression !== undefined) {
      return getNodeHandler(nodeAsJsxExpression.expression.kind).getData(
        nodeAsJsxExpression.expression,
        rootNode
      )
    } else {
      // return raw data of node
      return new DefaultHandler().getData(node, rootNode)
    }
  }
}
