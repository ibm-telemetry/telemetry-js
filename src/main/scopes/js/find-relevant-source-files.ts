/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import getPropertyByPath = require('lodash/get.js')
import * as path from 'node:path'

import { ObjectPath } from 'object-scan'

import { getTrackedSourceFiles } from '../../core/get-tracked-source-files.js'
import { Logger } from '../../core/log/logger.js'
import { NoInstallationFoundError } from '../../exceptions/no-installation-found-error.js'
import { getDependencyTree } from '../npm/get-dependency-tree.js'
import { getDirectoryPrefix } from '../npm/get-directory-prefix.js'
import { getInstalledVersionPaths } from '../npm/get-installed-version-paths.js'
import { getPackageData } from '../npm/get-package-data.js'
import { getPackageTrees } from '../npm/get-package-trees.js'
import { getTreePredecessor } from '../npm/get-tree-predecessor.js'
import { DependencyTree, PackageData } from '../npm/interfaces.js'

/**
 * Finds tracked source files and then filters them based on ones that appear in a  project which
 * depends on the in-context instrumented package/version.
 *
 * @param instrumentedPackage - Data about the instrumented package to use during filtering.
 * @param cwd - Current working directory. This must be inside of the root directory. This is an
 * absolute path.
 * @param root - Root-most directory. This is an absolute path.
 * @param fileExtensions - List of file extensions to capture metrics for.
 * @param logger - Logger instance.
 * @returns A (possibly empty) array of source files.
 */
export async function findRelevantSourceFiles(
  instrumentedPackage: PackageData,
  cwd: string,
  root: string,
  fileExtensions: string[],
  logger: Logger
) {
  logger.traceEnter('', 'findRelevantSourceFiles', [instrumentedPackage, cwd, root, fileExtensions])
  const sourceFiles = await getTrackedSourceFiles(root, logger, fileExtensions)

  const dependencyTree = await getDependencyTree(cwd, root, logger)

  const filterPromises = sourceFiles.map(async (f) => {
    const prefix = await getDirectoryPrefix(path.dirname(f.fileName), logger)
    const prefixPackageData = await getPackageData(prefix, root, logger)

    let packageTrees = getPackageTrees(dependencyTree, prefixPackageData)

    let instrumentedInstallVersions: string[] | undefined = undefined
    let shortestPathLength: number | undefined = undefined
    do {
      for (const tree of packageTrees) {
        const instrumentedInstallPaths = getInstalledVersionPaths(tree, instrumentedPackage.name)
        if (instrumentedInstallPaths.length > 0) {
          const pathsLength = instrumentedInstallPaths[0]?.length ?? 0
          if (shortestPathLength === undefined || pathsLength < shortestPathLength) {
            instrumentedInstallVersions = instrumentedInstallPaths.map(
              (path) => getPropertyByPath(tree, path)['version']
            )
            shortestPathLength = pathsLength
          }
        }
      }
      // did not find, go up one level for all packages
      packageTrees = packageTrees
        .map((tree) => getTreePredecessor(dependencyTree, tree['path'] as ObjectPath))
        .filter((tree) => tree !== undefined) as DependencyTree[]
    } while (shortestPathLength === undefined && packageTrees.length > 0)

    if (instrumentedInstallVersions === undefined) {
      throw new NoInstallationFoundError(instrumentedPackage.name)
    }

    return instrumentedInstallVersions.some((version) => version === instrumentedPackage.version)
  })
  const filterData = await Promise.all(filterPromises)

  const results = sourceFiles.filter((_, index) => {
    return filterData[index]
  })

  logger.traceEnter('', 'findRelevantSourceFiles', results)
  return results
}
