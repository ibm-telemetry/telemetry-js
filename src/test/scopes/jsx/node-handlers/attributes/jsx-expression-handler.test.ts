/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as ts from 'typescript'
import { describe, expect, it } from 'vitest'

import { NoAttributeExpressionFoundError } from '../../../../../main/exceptions/no-attribute-expression-found-error.js'
import { JsxExpressionHandler } from '../../../../../main/scopes/jsx/node-handlers/attributes/jsx-expression-handler.js'
import { getTrackedSourceFiles } from '../../../../../main/scopes/jsx/utils/get-tracked-source-files.js'
import { findNodesByType } from '../../../../__utils/find-nodes-by-type.js'
import { Fixture } from '../../../../__utils/fixture.js'
import { initLogger } from '../../../../__utils/init-logger.js'

describe('class: JsxExpressionHandler', () => {
  const logger = initLogger()
  it('correctly returns node text', async () => {
    const fixture = new Fixture('jsx-samples/all-attr-types.tsx')
    const sourceFile = (await getTrackedSourceFiles(fixture.path, logger))[0] as ts.SourceFile
    const handler = new JsxExpressionHandler(sourceFile, logger)

    const nodes = findNodesByType<ts.JsxExpression>(sourceFile, ts.SyntaxKind.JsxExpression)
    const data = nodes.map(handler.getData.bind(handler))

    expect(data).toMatchSnapshot()
  })

  it('throws NoAttributeExpressionFoundError if attribute does not have an expression', async () => {
    const fixture = new Fixture('jsx-samples/all-jsx-element-types.tsx')
    const sourceFile = (await getTrackedSourceFiles(fixture.path, logger))[0] as ts.SourceFile
    const handler = new JsxExpressionHandler(sourceFile, logger)

    const node = findNodesByType(
      sourceFile,
      ts.SyntaxKind.JsxExpression,
      (node) => (node as ts.JsxExpression).expression === undefined
    )[0]

    expect(() => handler.getData(node as ts.JsxExpression)).toThrow(NoAttributeExpressionFoundError)
  })
})
