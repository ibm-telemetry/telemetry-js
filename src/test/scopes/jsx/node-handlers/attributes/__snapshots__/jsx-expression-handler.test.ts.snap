// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`class: JsxExpressionHandler > correctly returns node text 1`] = `
[
  5,
  undefined,
  null,
  ComplexAttribute {
    "complexValue": "{key: 'value'}",
  },
  true,
  false,
  ComplexAttribute {
    "complexValue": "obj['accessKey']",
  },
  ComplexAttribute {
    "complexValue": "someIdentifier",
  },
  ComplexAttribute {
    "complexValue": "1 === 5 ? 'expression lhs' : 'expression rhs'",
  },
]
`;
